crates/sui-graphql-rpc/src/data/move_registry_data_loader.rs

```diff
@@ -0,0 +1,195 @@
+    }
+}
+
+impl MoveRegistryDataLoader {
+    pub(crate) fn new(config: MoveRegistryConfig) -> Self {
+        let batch_size = config.page_limit as usize;
+        let data_loader = DataLoader::new(ExternalNamesLoader::new(config), tokio::spawn)
+            .max_batch_size(batch_size);
+        Self(Arc::new(data_loader))
+    }
+}
+
+#[async_trait::async_trait]
+impl Loader<Name> for ExternalNamesLoader {
+    type Value = AppRecord;
+    type Error = Error;
+
+    /// This function queries the external API to fetch the app records for the requested names.
+    /// This is part of the data loader, so all queries are bulked-up to the maximum of {config.page_limit}.
+    /// We handle the cases where individual queries fail, to ensure that a failed query cannot affect
+    /// a successful one.
+    async fn load(&self, keys: &[Name]) -> Result<HashMap<Name, AppRecord>, Error> {
+        let Some(api_url) = self.config.external_api_url.as_ref() else {
+            return Err(Error::MoveNameRegistry(
+                MoveRegistryError::ExternalApiUrlUnavailable,
+            ));
+        };
+
+        let mut results: HashMap<Name, AppRecord> = HashMap::new();
+
+        let (query, mapping) = self.construct_names_graphql_query(keys);
+
+        let request_body = GraphQLRequest {
+            query,
+            variables: serde_json::Value::Null,
+        };
+
+        let res = self
+            .client
+            .post(api_url)
+            .json(&request_body)
+            .send()
+            .await
+            .map_err(|e| {
+                Error::MoveNameRegistry(MoveRegistryError::FailedToQueryExternalApi(e.to_string()))
+            })?;
+
+        if !res.status().is_success() {
+            return Err(Error::MoveNameRegistry(
+                MoveRegistryError::FailedToQueryExternalApi(format!(
+                    "Status code: {}",
+                    res.status()
+                )),
+            ));
+        }
+
+        let response_json: GraphQLResponse<Owner> = res.json().await.map_err(|e| {
+            Error::MoveNameRegistry(MoveRegistryError::FailedToParseExternalResponse(
+                e.to_string(),
+            ))
+        })?;
+
+        let names = response_json.data.owner.names;
+
+        for k in mapping.keys() {
+            // SAFETY: we inserted the keys in the mapping before.
+            let idx = mapping.get(k).unwrap();
+
+            let Some(Some(bcs)) = names.get(&fetch_key(idx)) else {
+                continue;
+            };
+
+            let Some(bytes) = Base64::from_str(&bcs.value.bcs).ok() else {
+                continue;
+            };
+
+            let Some(app_record) = bcs::from_bytes::<AppRecord>(&bytes.0).ok() else {
+                continue;
+            };
+
+            // only insert the record if it is a valid `app_record`
+            results.insert(k.clone(), app_record);
+        }
+
+        Ok(results)
+    }
+}
+
+impl Default for ExternalNamesLoader {
```

Comment:
Out of interest, what do you use this default impl for?

Reply:
Oh, it's not needed. I can't remember why I added it at some point. Removed it.